### PR TITLE
Convert Rvsdgs back into Cfgs

### PR DESCRIPTION
**Pull request by oflatt**
This PR adds a conversion from Rvsdg back to Cfg and tests it. The conversion works in a conservative way, and isn't good at re-using variables. It also doesn't implement reconstruction, so it always outputs a structured CFG for now.

It does not hook up the whole pipeline into egglog and back out. I'll be working on that next.
Merge #40 first.

Also disabled the `two_fns` test for now.

### CODE REVIEW

File: src/rvsdg/to_cfg.rs [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+                        .push(IncompleteBranch::Direct { from: branch_block });
+                }
+
+                shared_vars.unwrap()
+            }
+            RvsdgBody::Theta {
+                pred: pred_operand,
+                inputs,
+                outputs,
+            } => {
```
Comment (yihozhang) [inline src/rvsdg/to_cfg.rs:345]: I find this hard to understand. Some comments (or even ASCII art) for how the CFG layout for Theta node will be very helpful!

**Review by yihozhang — CHANGES_REQUESTED** [review]
Thanks for the PR! This is a very clever algorithm, with a translation scheme based on `IncompleteBranch` and `finish_block`. Some high-level comments:

* Is your PR tested (besides snapshot test)? It would be great to have some unit test for this conversion from RVSDG to CFG in particular, as well as "roundtrip test" to see the program produces the same output.
* I would really appreciate it if you can add some paragraph at the top of the `to_cfg.rs` file to explain how the algorithm works. This will help other team members to understand and contribute to this code. Right now one has to read through the entire file to understand how the algorithm works.
* I saw your PR has several unrelated changes to the codebase. I think this is fine right now since the codebase is not yet stable, but in near future we should consider making the codebase more stable and have each PR more dedicated to its purpose (without unrelated changes).

**Comment by oflatt** [comment]
> Thanks for the PR! This is a very clever algorithm, with a translation scheme based on `IncompleteBranch` and `finish_block`. Some high-level comments:
> 
> * Is your PR tested (besides snapshot test)? It would be great to have some unit test for this conversion from RVSDG to CFG in particular, as well as "roundtrip test" to see the program produces the same output after optimization.

Yep, this pr has a roundtrip test that makes sure we get the same output after optimization. I agree some unit tests would be nice but not high priority for me right now.

> * I would really appreciate it if you can add some paragraph at the top of the `to_cfg.rs` file to explain how the algorithm works. This will help other team members to understand and contribute to this code. Right now one has to read through the entire file to understand how the algorithm works.

Good idea, I'll add it.

> * I saw your PR has several unrelated changes to the codebase. I think this is fine right now since the codebase is not yet stable, but in near future we should consider making the codebase more stable and have each PR more dedicated to its purpose (without unrelated changes).

Good idea, in the future I'll do unrelated changes separately. I think I struggle with that in general haha

File: src/cfg/to_structured.rs [hunk 2]
```diff
@@ -68,6 +68,7 @@ impl<'a> StructuredCfgBuilder<'a> {
     /// to a structured representation.
     fn do_tree(&mut self, node: NodeIndex) -> StructuredBlock {
         if self.is_loop_header(node) {
+            eprintln!("loop header {:?}", self.name(node));
```
[the remark this answers is not in the document]
Comment (oflatt) [reply src/cfg/to_structured.rs]: yep thanks

File: src/rvsdg/to_cfg.rs [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+            exit: last_block,
+            return_ty: None,
+            _phantom: Simple,
+        }
+    }
+}
+
+impl<'a> RvsdgToCfg<'a> {
+    // Returns the name of the variable storing the result
+    // or None if no value is returned.
```
[the remark this answers is not in the document]
Comment (oflatt) [reply src/rvsdg/to_cfg.rs:187]: Removed good point

File: src/rvsdg/to_cfg.rs [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+
+        self.operand_cache.insert((*context, operand), res.clone());
+        res
+    }
+
+    // helper function to assigning to a set of variables
+    // this is helpful in looping for loop variables or assigning to shared
+    // variables across branches in a gamma
+    fn assign_to_vars(&mut self, input_vars: &[RvsdgValue], resulting_vars: &[RvsdgValue]) {
+        // assign to the loop variables, making sure the types line up
```
[the remark this answers is not in the document]
Comment (oflatt) [reply src/rvsdg/to_cfg.rs]: done

File: src/rvsdg/to_cfg.rs [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+        var: Identifier,
+        val: bool,
+    },
+    /// Direct jump from a given block to the next one
+    Direct { from: NodeIndex },
+}
+
+/// Represents the result of a RVSDG computation
+#[derive(Clone, Debug)]
+enum RvsdgValue {
```
[the remark this answers is not in the document]
Comment (oflatt) [reply src/rvsdg/to_cfg.rs:73]: Will rename after my PRs are in

File: src/rvsdg/to_cfg.rs [hunk 6]
```diff
@@ -0,0 +1,523 @@
...
+                    .collect::<Vec<_>>();
+
+                // evaluate pred in this block as well
+                // TODO we are assuming pred is an int here, is that actually true?
+                let pred = self.operand_to_bril(*pred, current_args, ctx);
+                let pred_bool = self.cast_bool(&pred);
+                let prev_block = self.finish_block();
+
+                let mut branch_blocks = vec![];
+                let mut shared_vars = None;
```
[the remark this answers is not in the document]
Comment (oflatt) [reply src/rvsdg/to_cfg.rs:331]: Wrote comment, didn't figure out how to make it more elegant

File: src/rvsdg/to_cfg.rs [hunk 7]
```diff
@@ -0,0 +1,523 @@
...
+            pos: None,
+        };
+        let res = self.graph.add_node(block);
+        if self.entry_block.is_none() {
+            self.entry_block = Some(res);
+        }
+
+        // drain the queue of incomplete branches
+        for branch in std::mem::take(&mut self.incomplete_branches) {
+            match branch {
```
[the remark this answers is not in the document]
Comment (oflatt) [reply src/rvsdg/to_cfg.rs]: Great idea

File: src/rvsdg/to_cfg.rs (hunk 1 above)
Comment (oflatt) [reply src/rvsdg/to_cfg.rs:393]: Done!

**Review by yihozhang — APPROVED** [review]
LGTM!

**Pull request merged**